File: Ruby/Hangar.rb
#encoding: utf-8

# +-------------------------------------------------------------------+ #
# | DEEPSPACE (Ruby)                                                  | #
# | Authors: Shao Jie Hu Chen and Mario Megías Mateo                  | #
# +-------------------------------------------------------------------+ #
# | This class specifies the characteristics of an Hangar.            | #
# +-------------------------------------------------------------------+ #

module Deepspace

    require_relative 'ShieldBooster.rb'
    require_relative 'Weapon.rb'
    require_relative 'HangarToUI.rb'

    class Hangar

        # Initialize
        def initialize(capacity)
            @maxElements    = capacity
            @shieldBoosters = Array.new
            @weapons        = Array.new
        end

        # Getters

        attr_reader :maxElements

        def shieldBoosters 
            Array.new(@shieldBoosters)
        end

        def weapons
            Array.new(@weapons)
        end

        # Copy constructor
        def self.newCopy(h)
            mod = new(h.maxElements)

            h.shieldBoosters.each do |s|
                mod.addShieldBooster(s)
            end

            h.weapons.each do |w|
                mod.addWeapon(w)
            end

            return mod
        end

        def getUIversion
            HangarToUI.new(self)
        end

        def spaceAvailable 
            shieldBoosters.length + weapons.length < maxElements
        end

        private :spaceAvailable 

        def addWeapon(w)
            if spaceAvailable()
                @weapons << Weapon.newCopy(w)
                return true
            else
                return false
            end
        end

        def addShieldBooster(s)
            if spaceAvailable()
                @shieldBoosters << ShieldBooster.newCopy(s)
                return true
            else
                return false
            end
        end

        def removeShieldBooster(s)
            if 0<=s && s<@shieldBoosters.length
                @shieldBoosters.delete_at(s)
            else
                return nil
            end
            
        end

        def removeWeapon(w)
            if 0<=w && w<@weapons.length
                @weapons.delete_at(w)
            else
                return nil
            end
        end

        def to_s

            cadShieldBoosters = ""
            @shieldBoosters.each do |shield|
                cadShieldBoosters += shield.to_s
            end
            cadWeapons = ""
            @weapons.each do |weapon|
                cadWeapons += weapon.to_s
            end

            s = "["
            s += "maxElements: " + @maxElements.to_s
            s += "; shieldBoosters: " + cadShieldBoosters
            s += "; weapons: " + cadWeapons
            s += "]"

            return s
        end

    end # end of Hangar
 
end # end of Deepspace

File: Java/src/deepspace/Hangar.java
/**
 * @brief DEEPSPACE (Java). Hangar
 * This class specifies the characteristics of an Hangar. 
 * 
 * @author Mario Megías Mateo and Shao Jie Hu Chen.
 */

package deepspace;
import java.util.ArrayList;

public class Hangar {
    private int maxElements;
    private ArrayList<Weapon> weapons;
    private ArrayList<ShieldBooster> shieldBoosters;

    /**
     * @brief Constructor
     * @param capacity Initial capacity of the plane. 
     */
    Hangar(int capacity) {
        maxElements = capacity;
        weapons = new ArrayList<Weapon>();
        shieldBoosters = new ArrayList<ShieldBooster>();
    }

    Hangar(Hangar one) {
        this.maxElements = one.maxElements;
        this.weapons = new ArrayList<Weapon>(one.weapons);
        this.shieldBoosters = new ArrayList<ShieldBooster>(one.shieldBoosters);
    }

    public HangarToUI getUIversion() {
        return new HangarToUI(this);
    }

    private boolean spaceAvailable() {
        return (this.weapons.size() + this.shieldBoosters.size()) < this.maxElements;
    }

    public boolean addWeapon(Weapon w) {
        if (this.spaceAvailable()) {
            this.weapons.add(new Weapon(w));
            return true;
        } else {
            return false;
        }
    }

    public boolean addShieldBooster(ShieldBooster s) {
        if (this.spaceAvailable()) {
            this.shieldBoosters.add(new ShieldBooster(s));
            return true;
        } else {
            return false;
        }
    }

    public int getMaxElements() {
        return this.maxElements;
    }

    public ArrayList<ShieldBooster> getShieldBoosters() {
        return new ArrayList<ShieldBooster>(this.shieldBoosters);
    }

    public ArrayList<Weapon> getWeapons() {
        return new ArrayList<Weapon>(this.weapons);
    }

    public ShieldBooster removeShieldBooster(int s) {
        if (0 <= s && s < this.shieldBoosters.size()) {
            ShieldBooster removedShieldBooster = new ShieldBooster(this.shieldBoosters.get(s));
            this.shieldBoosters.remove(s);

            return removedShieldBooster;
        } else {
            return null;
        }
    }

    public Weapon removeWeapon(int w) {
        if (0 <= w && w < this.weapons.size()) {
            Weapon removedWeapon = new Weapon(this.weapons.get(w));
            this.weapons.remove(w);

            return removedWeapon;
        } else {
            return null;
        }
    }

    public String toString() {
        String data = "";

        data += "Max Elements: " + this.maxElements + "\n";
        data += "Weapons: " + this.weapons.toString() + "\n";
        data += "ShieldBoosters: " + this.shieldBoosters.toString() + "\n";

        return data;
    }
}
